### ISSUE TITLE
Replace `HOST_c2l` in md32_common.h with byteswap function

### ISSUE THREAD

**Issue by briansmith** [briansmith/ring#71]
`HOST_c2l` modifies its argument, which isn't obvious when reading the code. Also, all the users are of this form:

```
HOST_c2l(data, l);
X(n) = l;
```

This can be simplified to:
```
X(n) = HOST_c2l(data);
data += 4;
```.

**Comment by frewsxcv** [comment]
Can I take this?

**Comment by frewsxcv** [comment]
sort of relevant: https://boringssl-review.googlesource.com/#/c/6652/

**Comment by briansmith** [comment]
Yes, go ahead. For some reason I can't assign the bug to you. Check out the definition of BSWAP4 in crypto/modes/internal.h. That should probably be moved to crypto/internal.h and reused everywhere.

**Comment by frewsxcv** [comment]
> For some reason I can't assign the bug to you.

Unfortunately: "You can only create assignments for yourself, collaborators on personal projects, or members of your organization with read permissions on the repository."

https://help.github.com/articles/assigning-issues-and-pull-requests-to-other-github-users/

**Comment by frewsxcv** [comment]
How does `BSWAP4` relate to `HOST_c2l`?

**Comment by briansmith** [comment]
> How does BSWAP4 relate to HOST_c2l?

Here is one of the definitions of `HOST_c2l`:

```
#define HOST_c2l(c, l)                       \
  (void)({                                   \
    uint32_t r = *((const uint32_t *)(c));   \
    __asm__("bswapl %0" : "=r"(r) : "0"(r)); \
    (c) += 4;                                \
    (l) = r;                                 \
  })
```

**Comment by frewsxcv** [comment]
@briansmith Did you mean for the second code block in your original post to be:

```c
X(n) = BSWAP4(data);
data += 4;
```

**Comment by briansmith** [comment]
Something like that. Note that these functions need to be defined as no-ops for big-endian systems, and bswap on little-endian systems. At the time I wrote the initial ones, I didn't think about HOST_c2l being the same as bswap on little-endian platforms. The Rust functions are called `to_be` so renaming these functions `to_be_u32` & `from_be_u32` for 32-bit and `to_be_u64` & `from_be_u64` for 64-bit. (Note: The 64-bit versions are in sha512.c.)

This might be something best split up into multiple patches.

Also, instead of using `#define`, please define them as `static inline` functions.

**Comment by briansmith** [comment]
Fixed by 16af088fbc2099d3b03e04b8086278cdf03104e1. There was a small (~5%) performance regression from the change that was addressed (at least partially) by 60eedb9105669400a8c902119b32ab4024e729a6. Also, there was some undefined behavior caused by shifting  into the sign bit (`x << 24` where `x` is a `uint8_t`) that was resolved by generous application of casting.

**Comment by briansmith** [comment]
Thanks for fixing this! I'm glad that md32_common.h is gone and that there's less assembly language code now.
